File: Implementations/Dining Philosophers/dining_philosophers.cpp
#include <iostream>
#include <vector>
#include <thread>
#include <mutex>
#include <unistd.h>
#include <chrono>

int constexpr NUM_PHILOSOPHERS = 5;
int constexpr TOTAL_BITES_EATEN = 3000;

struct Fork {
  int index;
  std::mutex mutex;

  Fork &operator=(const Fork &f) {
      index = f.index;
      return *this;
  }
};

class Philosopher {
  int bites_eaten = 0;

  public:
    int index = 0;
    Fork& dominant_hand;
    Fork& other_hand;

    Philosopher(int index1, Fork& dominant_hand1, Fork& other_hand1): dominant_hand(dominant_hand1), other_hand(other_hand1) {
      index = index1;
    }

    void think() {
      // usleep(2e+6);
      // std::cout << index << " is thinking" << std::endl;
    }

    void get_forks() {
      dominant_hand.mutex.lock();
      // std::cout << index << " is waiting for left fork" << std::endl;
      other_hand.mutex.lock();
      // std::cout << index << " is waiting for right fork" << std::endl;
    }

    void eat() {
      bites_eaten++;
      // std::cout << index << " has eaten " << bites_eaten << " times" << std::endl;
    }

    void put_forks() {
      dominant_hand.mutex.unlock();
      other_hand.mutex.unlock();
      // std::cout << index << " has put forks down" << std::endl;
    }

    bool is_done_eating() {
      return bites_eaten == TOTAL_BITES_EATEN;
    }
};

int main() {
  auto begin = std::chrono::steady_clock::now();

  // Create the forks for the philosophers to use
  std::vector<Fork> forks(NUM_PHILOSOPHERS);
  for (auto i = 0; i < NUM_PHILOSOPHERS; ++i) {
    Fork f;
    f.index = i;
    forks[i] = f;
  }

  std::vector<std::thread> threads;
  for (auto i = 0; i < NUM_PHILOSOPHERS; ++i) {
    Philosopher p(i, i == 0 ? forks[(i + 1) % NUM_PHILOSOPHERS] : forks[i], i == 0 ? forks[i] : forks[(i + 1) % NUM_PHILOSOPHERS]);

    threads.push_back(std::thread ([p] () mutable {
        while (!p.is_done_eating()) {
          p.think();
          p.get_forks();
          p.eat();
          p.put_forks();
        }
    }));
  }

  for (auto& t : threads) {
    t.join();
  }

  auto end = std::chrono::steady_clock::now();
  std::cout << "Time difference = " << std::chrono::duration_cast<std::chrono::microseconds>(end - begin).count() << "µs" << std::endl;
}

File: Implementations/Dining Philosophers/dining_philosophers.go
package main

import (
  "fmt"
  "sync"
  "time"
)

const numPhilosophers = 5
const totalBitesEaten = 3000;

type Fork struct {
  index int
  mutex sync.Mutex
}

type Philosopher struct {
  index int
  dominant_hand *Fork
  other_hand *Fork
  bites_eaten int
}

func (p *Philosopher) think() {
  // time.Sleep(2 * time.Second)
  // fmt.Printf("%v is thinking\n", p.index)
}

func (p *Philosopher) get_forks() {
  // fmt.Printf("%v is waiting for left fork\n", p.index)
  p.dominant_hand.mutex.Lock()
  // fmt.Printf("%v is waiting for right fork\n", p.index)
  p.other_hand.mutex.Lock()
}

func (p *Philosopher) eat() {
  p.bites_eaten++
  // fmt.Printf("%v has eaten %v times\n", p.index, p.bites_eaten)
}

func (p *Philosopher) put_forks() {
  p.dominant_hand.mutex.Unlock()
  p.other_hand.mutex.Unlock()
  // fmt.Printf("%v has put forks down\n", p.index)
}

func (p *Philosopher) is_done_eating(dining *sync.WaitGroup) bool {
  if p.bites_eaten == totalBitesEaten {
    dining.Done()
  }

  return p.bites_eaten == totalBitesEaten
}

func main() {
  start := time.Now()
  var dining sync.WaitGroup

  // Create the forks for the philosophers to use
  forks := make([]Fork, 0, numPhilosophers)
  for i := 0; i < numPhilosophers; i++ {
    forks = append(forks, Fork{index: i})
  }

  for i := 0; i < numPhilosophers; i++ {
    dining.Add(1)
    p := Philosopher{i, &forks[i], &(forks[(i + 1) % numPhilosophers]), 0}
    if i == 0 {
      // Make one philosopher left handed
      p = Philosopher{i, &(forks[(i + 1) % numPhilosophers]), &forks[i], 0}
    }

    go func(p *Philosopher, dining *sync.WaitGroup) {
      for !p.is_done_eating(dining) {
        p.think()
        p.get_forks()
        p.eat()
        p.put_forks()
      }
    }(&p, &dining)
  }

  dining.Wait()
  fmt.Printf("Time difference = %s\n", time.Since(start))
}
